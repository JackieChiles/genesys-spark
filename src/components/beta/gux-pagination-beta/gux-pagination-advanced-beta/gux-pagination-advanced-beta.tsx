import {
  Component,
  ComponentInterface,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  State
} from '@stencil/core';

import { trackComponent } from '../../../../usage-tracking';

import {
  GuxItemsPerPage,
  GuxPaginationState
} from '../gux-pagination-beta.types';

import {
  calculateTotalPages,
  calculateCurrentPage
} from '../gux-pagination-beta.service';

@Component({
  styleUrl: '../gux-pagination-beta.less',
  tag: 'gux-pagination-advanced-beta',
  shadow: true
})
export class GuxPaginationAdvancedBeta implements ComponentInterface {
  @Element()
  private root: HTMLElement;

  /**
   * The currently select page. Changes are watched by the component.
   */
  @Prop({ mutable: true })
  currentPage: number = 1;

  /**
   * The total number of items in the data set. Used to calculate total page count
   */
  @Prop()
  totalItems: number = 0;

  /**
   * The max number of items on a page. Used to calculate total page count
   */
  @Prop({ mutable: true })
  itemsPerPage: GuxItemsPerPage = 25;

  /**
   * True when the total number of pages is unknown.
   */
  @Prop()
  pagesUnknown: boolean = false;

  /**
   * The total number of pages needed for the the data set.
   */
  @State()
  private totalPages: number;

  @Event()
  private guxpaginationchange: EventEmitter<GuxPaginationState>;

  private setPage(page: number): void {
    if (page < 0) {
      if (this.totalPages > 0) {
        this.setPage(1);
      } else {
        this.setPage(0);
      }

      return;
    }

    const totalPages = calculateTotalPages(this.totalItems, this.itemsPerPage);
    if (page > totalPages) {
      this.setPage(totalPages);
      return;
    }

    this.currentPage = page;
    this.guxpaginationchange.emit({
      currentPage: this.currentPage,
      itemsPerPage: this.itemsPerPage
    });
  }

  private handleInternalitemsperpagechange(event: CustomEvent): void {
    this.itemsPerPage = event.detail as GuxItemsPerPage;
    this.setPage(1);
  }

  private handleInternalcurrentpagechange(event: CustomEvent): void {
    this.setPage(event.detail as number);
  }

  componentWillLoad(): void {
    trackComponent(this.root, { variant: 'advanced' });
  }

  componentWillRender(): void {
    this.totalPages = calculateTotalPages(this.totalItems, this.itemsPerPage);
    this.currentPage = calculateCurrentPage(this.totalPages, this.currentPage);
  }

  render(): JSX.Element {
    return (
      <div class="gux-pagination-container">
        <div class="gux-pagination-info">
          <gux-pagination-item-counts-beta
            total-items={this.totalItems}
            current-page={this.currentPage}
            items-per-page={this.itemsPerPage}
            pages-unknown={this.pagesUnknown}
          />
          <gux-pagination-items-per-page-beta
            items-per-page={this.itemsPerPage}
            onInternalitemsperpagechange={this.handleInternalitemsperpagechange.bind(
              this
            )}
          />
        </div>
        <div class="gux-pagination-change">
          <gux-pagination-buttons-beta
            layout="advanced"
            current-page={this.currentPage}
            total-pages={this.totalPages}
            onInternalcurrentpagechange={this.handleInternalcurrentpagechange.bind(
              this
            )}
          />
        </div>
      </div>
    ) as JSX.Element;
  }
}
