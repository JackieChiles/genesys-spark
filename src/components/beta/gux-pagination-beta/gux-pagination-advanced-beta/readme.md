# gux-pagination-advanced-beta



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute        | Description                                                                   | Type                    | Default |
| -------------- | ---------------- | ----------------------------------------------------------------------------- | ----------------------- | ------- |
| `currentPage`  | `current-page`   | The currently select page. Changes are watched by the component.              | `number`                | `1`     |
| `itemsPerPage` | `items-per-page` | The max number of items on a page. Used to calculate total page count         | `100 \| 25 \| 50 \| 75` | `25`    |
| `pagesUnknown` | `pages-unknown`  | True when the total number of pages is unknown.                               | `boolean`               | `false` |
| `totalItems`   | `total-items`    | The total number of items in the data set. Used to calculate total page count | `number`                | `0`     |


## Events

| Event                 | Description | Type                                                          |
| --------------------- | ----------- | ------------------------------------------------------------- |
| `guxpaginationchange` |             | `CustomEvent<{ currentPage: number; itemsPerPage: number; }>` |


## Dependencies

### Used by

 - [gux-pagination-beta](..)

### Depends on

- [gux-pagination-item-counts-beta](../gux-pagination-item-counts-beta)
- [gux-pagination-items-per-page-beta](../gux-pagination-items-per-page-beta)
- [gux-pagination-buttons-beta](../gux-pagination-buttons-beta)

### Graph
```mermaid
graph TD;
  gux-pagination-advanced-beta --> gux-pagination-item-counts-beta
  gux-pagination-advanced-beta --> gux-pagination-items-per-page-beta
  gux-pagination-advanced-beta --> gux-pagination-buttons-beta
  gux-pagination-items-per-page-beta --> gux-dropdown-v2-beta
  gux-pagination-items-per-page-beta --> gux-listbox
  gux-pagination-items-per-page-beta --> gux-option-v2
  gux-dropdown-v2-beta --> gux-icon
  gux-dropdown-v2-beta --> gux-popup-beta
  gux-pagination-buttons-beta --> gux-button-slot-beta
  gux-pagination-buttons-beta --> gux-icon
  gux-pagination-beta --> gux-pagination-advanced-beta
  style gux-pagination-advanced-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
