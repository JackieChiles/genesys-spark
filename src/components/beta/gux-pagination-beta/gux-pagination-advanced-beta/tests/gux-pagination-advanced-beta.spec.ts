import { newSpecPage } from '@stencil/core/testing';
import { GuxPaginationAdvancedBeta } from '../gux-pagination-advanced-beta';

describe('gux-pagination-advanced-beta', () => {
  let page: SpecPage;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [GuxPaginationAdvancedBeta],
      html: `
      <gux-pagination-advanced-beta>
        total-items=123
        current-page=1
        items-per-page=25
        pages-unknown=false
        total-pages=4
      </gux-pagination-advanced-beta>
      `,
      language: 'en'
    });
  });

  it('should build', async () => {
    expect(page.rootInstance).toBeInstanceOf(GuxPaginationAdvancedBeta);
  });

  it('should render', async () => {
    expect(page.root).toMatchSnapshot();
  });
});
