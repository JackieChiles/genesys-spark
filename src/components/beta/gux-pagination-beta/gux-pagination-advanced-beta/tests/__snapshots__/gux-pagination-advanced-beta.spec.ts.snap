// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-pagination-advanced-beta should render 1`] = `
<gux-pagination-advanced-beta>
  <mock:shadow-root>
    <div class="gux-pagination-container">
      <div class="gux-pagination-info">
        <gux-pagination-item-counts-beta current-page="0" items-per-page="25" total-items="0"></gux-pagination-item-counts-beta>
        <gux-pagination-items-per-page-beta items-per-page="25"></gux-pagination-items-per-page-beta>
      </div>
      <div class="gux-pagination-change">
        <gux-pagination-buttons-beta current-page="0" layout="advanced" total-pages="0"></gux-pagination-buttons-beta>
      </div>
    </div>
  </mock:shadow-root>
  total-items=123 current-page=1 items-per-page=25 pages-unknown=false total-pages=4
</gux-pagination-advanced-beta>
`;
